src/main/java/at/ac/tuwien/kr/alpha/common/Assignment.java

```diff
@@ -11,7 +11,7 @@
 import static at.ac.tuwien.kr.alpha.solver.ThriceTruth.FALSE;
 import static at.ac.tuwien.kr.alpha.solver.ThriceTruth.TRUE;
 
-public interface Assignment {
+public interface Assignment<T extends Assignment.Entry> extends Iterable<T> {
```

Comment:
What is the reason for making `Assignment` parametrizable?
Now we get warnings in every place where `Assignment` is used as a raw type, so I suggest to either take this back or to add a type argument wherever `Assignment` is referenced.

Reply:
Since the implementation of `NaiveGrounder#getSomeMBTAssignedAtom` wants to iterate over the assignment, I thought that implementing `Iterable` would be reasonable. Iterating over an assignment seems quite a natural request. However, since `Assignment.Entry` is just an interface, we need generics to hint what the concrete implementation of `Assignment.Entry` will be.
I missed those warnings and added wildcards where the raw type was referenced.

What is a little ugly is that now we have `Set<Integer> Assignment#getTrueAssignments` in addition to `Assignment#iterator()`. Removing `#getTrueAssignments` in favor of `#iterator()` is not straight forward since `NaiveSolver` would then need some implementation of `Assignment`. The growing difference between `NaiveSolver` and `DefaultSolver` is indeed having a bad influence on our architecture. We might split them more or even drop `NaiveSolver` at some point.

Reply:
Actually, the reason it wants to iterate the assignment should be an implementation detail of the assignment. If there were a collection of pointers to MBT-assigned atoms, it would be much more economical to just iterate that and not the whole assigment. If the assignment would somehow know which atom ids represent internal atoms and which represent atoms over the domain of the input ASP program, this method would not be part of the grounder at all.

The problem with the justification is that in some sense it reverses our information flow: previously, the grounder was given assignments and deriving new ground instances; now the solver starts looking at rules itself and by that may also arrive at unseen ground instances. I (partially) solved this issue by moving the justification analysis into the grounder and handing it the current assignment. Maybe a different approach would work better? But I am not seeing it yet. Having the justification analysis in the solver created even more problems because of dependencies on grounder-internal knowledge.